PhIPSeq_external/Analysis_allergens_and_IEDB/prediction_helpers.py

```diff
@@ -0,0 +1,233 @@
+    params['max_depth'] = tune_single_hyper_parameter(x, y, is_classifier, 'max_depth', range(1, 201), **params)
+    params['min_samples_leaf'] = tune_single_hyper_parameter(x, y, is_classifier, 'min_samples_leaf', range(1, 201),
+                                                             **params)
+    return params
+
+
+def draw_auc(x, y, is_classifier, ax=None, num_iterations=0, color='blue', **model_params):
+    model_params['random_state'] = 46541357
+    df = predict_with_cross_validation(x, y, is_classifier, return_predictions=True, **model_params)
+    fpr, tpr, thresholds = roc_curve(df['y'], df['y_hat'])
+    label = f"Predictor (AUC={auc(fpr, tpr):.3f}"
+    if ax is None:
+        ax = plt.subplot()
+    if num_iterations > 0:
+        auc_confidence_interval = []
+        for _ in range(num_iterations):
+            round_fprs, round_tprs, round_thresholds = roc_curve(
+                *list(map(lambda item: item[1], train_test_split(df)[0].iteritems())))
+            round_auc = auc(round_fprs, round_tprs)
+            auc_confidence_interval.append(round_auc)
+            ax.plot(round_fprs, round_tprs, color='grey', alpha=5 / num_iterations)
+        label += f", std={np.std(auc_confidence_interval):.3f}"
+    label += ")"
+
+    ax.plot(fpr, tpr, color=color,
+            label=label,
+            lw=2,
+            alpha=0.8, )
+    ax.plot([0, 1], [0, 1], linestyle="--", lw=2, color="r", label="Chance", alpha=0.8)
+    ax.legend()
+    ax.set_xlabel('False positive rate')
+    ax.set_ylabel('True positive rate')
+    return ax
+
+
+def plot_shap(x, y, is_classifier, params, target_name, specs=None, fig=None, overwrite=False):
+    if fig is None:
+        fig = plt.figure()
+    if specs is None:
+        specs = fig.add_gridspec(1, 1)
+    external_ax = fig.add_subplot(specs)
+    external_ax.axis('off')
+    inner_spec = specs.subgridspec(1, 2, width_ratios=[1, 4], wspace=0)
+    ax = fig.add_subplot(inner_spec[1])
+
+    output_path_prefix = os.path.join(out_path, '_'.join([target_name, 'shap']))
+    output_csv_path = output_path_prefix + '.csv'
+    output_pickle_path = output_path_prefix + '.pkl'
+    os.makedirs(os.path.dirname(output_path_prefix), exist_ok=True)
+    if overwrite or not os.path.exists(output_pickle_path):
+        model_class = GradientBoostingClassifier if is_classifier else GradientBoostingRegressor
+        clf = model_class(**params).fit(x, y)
+        explainer = shap.TreeExplainer(clf)
+        shap_values = explainer(x)
+        with open(output_pickle_path, 'wb') as f:
+            pickle.dump(shap_values, f)
+    with open(output_pickle_path, 'rb') as f:
+        shap_values = pickle.load(f)
+
+    # Save absolute Shapley values to table
+    if overwrite or not os.path.exists(output_csv_path):
+        shap_abs_values = pd.Series(data=np.abs(explainer.shap_values(x)).sum(axis=0),
+                                    index=x.columns).sort_values(ascending=False)
```

Comment:
explainer does not exist if os.path.exists(output_pickle_path) was already there and read.
this all needs to be together under the if in line 155

Reply:
this is incase a single file was created and there are issues before creating the second one. They should not be merged together.